From d4c7f1ddb2e5f5bccd27585a606f57cea6f0487c Mon Sep 17 00:00:00 2001
From: "River, Li" <river.li@intel.com>
Date: Tue, 12 Apr 2016 16:15:59 +0800
Subject: [PATCH 1/3] FOR_UPSTREAM [VPG]: drm/i915: [CHT_CR] Fix
 drrs->connector->panel.target_mode NULL pointer issue.

       "drrs->connector->panel.target_mode" may be cleared when
       it is used, so we should lock it when w/r it.

Change-Id: Idb18eeca2a083e7065b6ae3161003891e3b3b11e
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-26176
Reviewed-on: https://android.intel.com:443/493399
---
 drivers/gpu/drm/i915/intel_drrs.c     | 6 +++++-
 drivers/gpu/drm/i915/intel_dsi_drrs.c | 2 ++
 2 files changed, 7 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/i915/intel_drrs.c b/drivers/gpu/drm/i915/intel_drrs.c
index dd07148..11fb672 100644
--- a/drivers/gpu/drm/i915/intel_drrs.c
+++ b/drivers/gpu/drm/i915/intel_drrs.c
@@ -164,7 +164,11 @@ static void intel_cancel_idleness_drrs_work(struct i915_drrs *drrs)
 		return;
 
 	cancel_delayed_work_sync(&drrs->idleness_drrs_work->work);
+
+	/* Avoid target_mode is NULL when it is in used */
+	mutex_lock(&drrs->drrs_mutex);
 	drrs->connector->panel.target_mode = NULL;
+	mutex_unlock(&drrs->drrs_mutex);
 }
 
 static void intel_enable_idleness_drrs(struct i915_drrs *drrs)
@@ -362,6 +366,7 @@ int intel_drrs_init(struct drm_device *dev,
 		goto err_out;
 	}
 
+	mutex_init(&drrs->drrs_mutex);
 	if (!drrs->encoder_ops->init || !drrs->encoder_ops->exit ||
 					!drrs->encoder_ops->set_drrs_state) {
 		DRM_DEBUG("Essential func ptrs are NULL\n");
@@ -383,7 +388,6 @@ int intel_drrs_init(struct drm_device *dev,
 
 	/* SEAMLESS DRRS is supported and downclock mode also exist */
 	drrs->has_drrs = true;
-	mutex_init(&drrs->drrs_mutex);
 	drrs->drrs_state.current_rr_type = DRRS_HIGH_RR;
 	DRM_INFO("SEAMLESS DRRS supported on this panel.\n");
 
diff --git a/drivers/gpu/drm/i915/intel_dsi_drrs.c b/drivers/gpu/drm/i915/intel_dsi_drrs.c
index 9f00d4f..91f35e6 100644
--- a/drivers/gpu/drm/i915/intel_dsi_drrs.c
+++ b/drivers/gpu/drm/i915/intel_dsi_drrs.c
@@ -258,7 +258,9 @@ int intel_dsi_drrs_init(struct i915_drrs *drrs,
 	DRM_DEBUG("downclock_mode :\n");
 	drm_mode_debug_printmodeline(downclock_mode);
 
+	mutex_lock(&drrs->drrs_mutex);
 	panel->target_mode = NULL;
+	mutex_unlock(&drrs->drrs_mutex);
 
 	if (IS_VALLEYVIEW(intel_encoder->base.dev)) {
 
-- 
1.9.1

